### DISCUSSION TITLE
JSON rule trigger websend#Data=Done doesnt work when a rule executes websend

### DISCUSSION THREAD

**Discussion by satmanuk2** [arendst/Tasmota#16359] [Q&A] [answered]
Hi All,

i have a really frustrating problem with catching websend success/failures that i need some help with.

I have basic test rule = rule2 ON WebSend#Data=Done do status 0 endon

this rule trigger works perfectly if i execute the websend command from the console myself

17:47:34.893 CMD: WebSend [192.168.1.55] POWER4 Toggle endon
17:47:34.942 RSL: RESULT = {"WebSend":"Done"}
17:47:34.958 RUL: WEBSEND#DATA=DONE performs "status 0"

but it does NOT work when another rule executed the websend 
17:40:54.805 RUL: SSERIALRECEIVED#DATA=0X50DF55 performs "WebSend [192.168.1.55] POWER4 Toggle"
17:40:54.972 RSL: RESULT = {"WebSend":"Done"}

The JSON result payloads are identical, so i really cant figure out why one triggers happily and the other does not.

any ideas ? (version is 12.1.1)

thanks in advance

**Comment by barbudor** [comment, marked as answer]
Try with adding a Backlog before the command : `ON WebSend#Data=Done do Backlog status 0 endon`

**Reply by satmanuk2** [reply]
thanks for the reply,

just tried that, same problem

18:05:20.470 RUL: Stored uncompressed, would compress from 46 to 34 (-26%)
18:05:20.473 RSL: RESULT = {"Rule2":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":46,"Free":465,"Rules":"ON WebSend#Data=Done do Backlog status 0 endon"}}
18:05:26.295 RSL: RESULT = {"SSerialReceived":"0x50DF55"}
18:05:26.310 RUL: SSERIALRECEIVED#DATA=0X50DF55 performs "WebSend [192.168.1.55] POWER4 Toggle"
18:05:26.398 RSL: RESULT = {"WebSend":"Done"}

**Reply by satmanuk2** [reply]
it is also worth noting that it doesnt work in the same ruleset either (which i expect is less likely to work anyway, hence going straight for 2 rules)

rule1 ON SSerialReceived#Data=0x50DF55 do WebSend [192.168.1.55] POWER4 Toggle endon ON WebSend#Data=Done do Backlog status 0 endon

**Reply by barbudor** [reply]
Sorry my mistake, I meant the WebSend should be in Backlog, not the Status 0
My bad

**Reply by satmanuk2** [reply]
perfect.. that has sorted it... thank you

such a simple thing has been tripping me up all afternoon...  i guess in production i would have probably used Backlog for multiple commands anyway, but not when testing a single command... is it advisable to always use backlog regardless  ?

18:24:38.372 RUL: SSERIALRECEIVED#DATA=0X50DF55 performs "Backlog WebSend [192.168.1.55] POWER4 Toggle"
18:24:38.597 RSL: RESULT = {"WebSend":"Done"}
18:24:38.615 RUL: WEBSEND#DATA=DONE performs "Backlog status 0"

**Reply by barbudor** [reply]
Backlog helps when you want the output of a command executed in a rule to be processed by another rule.
A typical usage is using that with boot-up rule such to execute a command to get some parameter and then execute something the with output of that command.

**Reply by satmanuk2** [reply]
gotcha... 

given my lack of knowledge on that front, i should ask if im heading in the right direction altogether, or if there is a simpler way.

essentially i need communications fallback between 2 standalone tasmota devices (no broker/HA node involved)

the prefered method is WIFI (websend) however if a set ammount of time has passed without a successful response from websend, i then need to fallback to a secondary coms method & repeat the command (e.g. via RF).  N.B. the 'failed' response from websend takes a little too long, about 5 seconds, hence catching a lack of 'success' within a smaller time span.

is a ruleset the right way to go, or is there some fallback mechanism already built into DO or Websend error catching that could be utilised instead ?

**Reply by sfromis** [reply]
For reliable communication, MQTT the preferred method, the Tasmota support for http is not designed for reliability. Certainly no inbuilt fallback or error recovery.

Generally, if you need to be sure that the other end has gotten and acted on a message, you should have it return a signal about success.

**Reply by barbudor** [reply]
In themselves HTTP or MQTT have the same level of reliabilty as they are both based on TCP
MQTT introduce an additional element (the broker) which somehow adds a possible single point of failure. Also sending a message through the broker doesn't mean that it will be received by a device.

The point raised by sfromis on getting an acknoledge is key.
With HTTP as it is, the "Done" you get just confirm that WebSend has pushed the command to something that accepted a connection on port 80, not that the command as been effectively executed.
Similarly, Publish would not raise an error if the receiving end is away. The role of Publish is just to push the message to the broker.

In both case you need an acknowledge.

With MQTT you would get that by subscribing on the stat/theotherdevice/POWER topic to get the confirmation that the power has switched.
With HTTP, you can self-compile with  USE_WEBSEND_RESPONSE so that WebSend result include the feedback from the other side that can be processed in a rule.

**Reply by satmanuk2** [reply]
i know MQTT should be the preferred method, however the situation calls for fewer devices as possible.

i will implement the remote device feedback/handshake method you've suggested, rather than trying to infer failure by lack of websend response.

thanks again for your help, much appreciated.

**Reply by satmanuk2** [reply]
i will also do a compile with the USE_WEBSEND_RESPONSE and try that method as well.

thanks all.
